proxyd/consensus_aware.go

```diff
@@ -0,0 +1,392 @@
+package proxyd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/go-redis/redis/v8"
+)
+
+const (
+	PollerInterval = 1 * time.Second
+)
+
+type backendState struct {
+	backendStateMux sync.Mutex
+
+	latestBlockNumber string
+	latestBlockHash   string
+
+	lastUpdate time.Time
+
+	bannedUntil time.Time
+}
+
+type ConsensusPoller struct {
+	cancelFunc context.CancelFunc
+
+	BackendGroup      *BackendGroup
+	backendState      map[*Backend]*backendState
+	consensusGroupMux sync.Mutex
+	ConsensusGroup    []*Backend
+
+	Tracker      ConsensusTracker
+	asyncHandler ConsensusAsyncHandler
+}
+
+type ConsensusTracker interface {
```

Comment:
[nit] Would move the `ConsensusTracker` interface + implementation into its own file for readability.